src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -37,6 +46,270 @@
 
 static char * PartitionBound(Oid partitionId);
 static Relation try_relation_open_nolock(Oid relationId);
+static List * CreateFixPartitionConstraintsTaskList(Oid relationId);
+static List * WorkerFixPartitionConstraintCommandList(Oid relationId, uint64 shardId,
+													  List *checkConstraintList);
+static List * CheckConstraintNameListForRelation(Oid relationId);
+static bool RelationHasConstraint(Oid relationId, char *constraintName);
+static char * RenameConstraintCommand(Oid relationId, char *constraintName,
+									  char *shardIdAppendedConstraintName);
+
+
+PG_FUNCTION_INFO_V1(fix_partition_constraints);
+PG_FUNCTION_INFO_V1(worker_fix_partition_constraints);
+
+
+/*
+ * fix_partition_constraints fixes the constraint names of partitioned table shards on
+ * workers.
+ *
+ * Constraint names for partitioned table shards should have shardId suffixes if and only
+ * if they are unique or foreign key constraints. We mistakenly appended shardIds to
+ * constraint names on ALTER TABLE dist_part_table ADD CONSTRAINT .. queries prior to
+ * Citus 10. fix_partition_constraints determines if this is the case, and renames
+ * constraints back to their original names on shards.
+ */
+Datum
+fix_partition_constraints(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	EnsureCoordinator();
+
+	if (!PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("could not fix partition constraints: "
+							   "relation does not exist or is not partitioned")));
+	}
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("fix_partition_constraints can only be called "
+							   "for distributed partitioned tables")));
+	}
+
+	List *taskList = CreateFixPartitionConstraintsTaskList(relationId);
+	bool localExecutionSupported = true;
+	ExecuteUtilityTaskList(taskList, localExecutionSupported);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * worker_fix_partition_constraints fixes the constraint names on a worker given a shell
+ * table name and shard id.
+ */
+Datum
+worker_fix_partition_constraints(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	int32 shardId = PG_GETARG_INT32(1);
+	text *constraintNameText = PG_GETARG_TEXT_P(2);
+
+	if (!PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("could not fix partition constraints: "
+							   "relation does not exist or is not partitioned")));
+	}
+
+	char *constraintName = text_to_cstring(constraintNameText);
+	char *shardIdAppendedConstraintName = pstrdup(constraintName);
+	AppendShardIdToName(&shardIdAppendedConstraintName, shardId);
+
+	/* if shardId was appended to the constraint name, rename back to original */
+	if (RelationHasConstraint(relationId, shardIdAppendedConstraintName))
+	{
+		char *renameConstraintDDLCommand =
+			RenameConstraintCommand(relationId, constraintName,
+									shardIdAppendedConstraintName);
```

Comment:
are the arguments swapped here?

the current logic seems to be: if X exists, rename Y to X.

Reply:
thanks for letting me know. I spent too much time because of this dumb mistake